Redesigning the workflow showing method

sagar15795 opened the pull request:
**Workflow List** [Done]
- [x]  Design UI for WorkFlows.
- [x]  Add Retrofit Network Request in ApiManager 
- [x]  Add ApiManger wiring with the DataManager.
- [x]  Add Network call in WorkFlows Presenter (Logic class) class and showing in RecyclerView.

**Detail Workflow** [WIP]

- [x] Design UI for Detail WorkFlow.
- [x] Add Retrofit Network Request in ApiManager for Detail Workflow
- [x] Add ApiManger wiring with the DataManager for Detail Workflow.
- [x] Add Network call in DetailWorkFlow Presenter (Logic class) class and show it in UI .
- [x] Add favourite FAB with working 
- [x] Add Workflow's Licence in overflow menu

Add zoom functionality  to  workflow image [will add it in next PR]

therajanmaurya commented:
@sagar15795 your build is failing. Please before push, run Gradle check.

sagar15795 commented:
Hi @therajanmaurya & folks

Have a look on Workflow UI 

![device-2016-06-23-020001](https://cloud.githubusercontent.com/assets/5468111/16282448/6530662a-38e6-11e6-8b1e-de839032266b.png)

In Workflow list I am showing only 4 elements 

1. Title
2. Type
3. Date of Creation
4. Author name 

Tell me if there is any change or suggestion.

ianwdunlop commented:
Hello

In the date time field I would not show the hours and minutes only the
year/month/day. Are you going to show the author image or only show that on
the individual workflow page?

Cheers

Ian
On 22 Jun 2016 21:51, "Sagar" <notifications@github.com> wrote:

Hi @therajanmaurya <https://github.com/therajanmaurya> , @stain
<https://github.com/stain> , @ianwdunlop <https://github.com/ianwdunlop> &
folks

Have a look on Workflow UI

[image: device-2016-06-23-020001]
<https://cloud.githubusercontent.com/assets/5468111/16282448/6530662a-38e6-11e6-8b1e-de839032266b.png>

In Workflow list I am showing only 4 elements

   1. Title
   2. Type
   3. Date of Creation
   4. Author name

Tell me if there is any change or suggestion.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/incubator-taverna-mobile/pull/18#issuecomment-227869159>,
or mute the thread
<https://github.com/notifications/unsubscribe/AAjDdXOYgr1LpttJnFsUeQrdlPPBKZu8ks5qOaBdgaJpZM4I3u1M>
.

therajanmaurya commented:
@ianwdunlop 
No, I do not think to add the Author Image because Its need one more API call.

@sagar15795 replace the Time into year/month/day.

@ianwdunlop @stain I want to talk about the REST API design.
As I have worked on Many REST API. There is first API Call give sufficient data to inflate the View.
If we will add more files in workflows REST API http://www.myexperiment.org/workflows.xml?page=1.

Then We do not need to hit the server 25 times for every workflow again to get details of the workflow.

Thanks

ianwdunlop commented:
Hello

What extra details are needed in the workflow list? The myexperiment API is
not something we really control. The code is open so we could change it
(it's a ruby on rails project) but getting it deployed might be tricky.
@stian @alan do you have any influence here?

Cheers

Ian
On 23 Jun 2016 10:56, "Rajan Maurya" <notifications@github.com> wrote:

> @ianwdunlop <https://github.com/ianwdunlop>
> No, I do not think to add the Author Image because Its need one more API
> call.
>
> @sagar15795 <https://github.com/sagar15795> replace the Time into
> year/month/day.
>
> @ianwdunlop <https://github.com/ianwdunlop> @stain
> <https://github.com/stain> I want to talk about the REST API design.
> As I have worked on Many REST API. There is first API Call give sufficient
> data to inflate the View.
> If we will add more files in workflows REST API
> http://www.myexperiment.org/workflows.xml?page=1.
>
> Then We do not need to hit the server 25 times for every workflow again to
> get details of the workflow.
>
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-taverna-mobile/pull/18#issuecomment-228004031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAjDdQTMdYMztO0UjF42O5XezEtqsHvcks5qOlhogaJpZM4I3u1M>
> .
>

therajanmaurya commented:
We can add more fields in workflows list
1. Type
2. Uploader
3. Preview (Image Preview)
here is the workflows list http://www.myexperiment.org/workflows.xml?page=1
Here is the workflow API : http://www.myexperiment.org/workflow.xml?id=4

stain commented:
Yes, any myExperiment pull requests welcome, and @fbacall can review them.

I sent the links to the relevant code in https://github.com/myExperiment/myExperiment/tree/master/app in the [other email](https://lists.apache.org/thread.html/4e5f406d9fa3c5cbe2dc21ed19cc8e29390e99e4a1f940ae1de14ec7@%3Cdev.taverna.apache.org%3E) 

However I am not (really) a Rails developer, so I am not sure how it can be modified to provide more details in the `workflows.xml` result, perhaps @fbacall or @stuzart has some tips?

fbacall commented:
Hello.

There is a parameter that lets you display extra elements in the various "index" calls in the myExperiment API, described here:
http://wiki.myexperiment.org/index.php/Developer:API#Element_Selection

the list of elements available is here:
http://wiki.myexperiment.org/index.php/Developer:WorkflowsResource#Elements_for_workflow

so to show the type, uploader and preview, you would do:
http://www.myexperiment.org/workflows.xml?page=1&elements=title,type,uploader,preview

therajanmaurya commented:
Thanks, @fbacall . It's solved everything.
So, @sagar15795 complete workflows list. Then we will discuss Details about workflow layout and API calls.

therajanmaurya commented:
Hi, @fbacall.
I want to ask http://wiki.myexperiment.org/index.php/Developer:API REST API will be available in JSON response.

sagar15795 commented:
@therajanmaurya  

I have completed Workflows. Everything working properly. 
Please Review it.
Tell me if there is any suggestion or error.

I will add licence after your review.

stain commented:
@sagar15795 

The [travis build](https://travis-ci.org/apache/incubator-taverna-mobile/builds/140005253) gives errors:

```
Execution failed for task ':app:checkstyle'.
> Checkstyle rule violations were found. See the report at: file:///home/travis/build/apache/incubator-taverna-mobile/app/build/reports/checkstyle/checkstyle.html
```

You need to add ASF headers in the new `model` bean classes.

stain commented:
@sagar15795 where did you find `placeholder.png`? What is its license and copyright?

stain commented on app/build.gradle:
```diff
@@ -59,4 +61,7 @@ dependencies {
     }
+
+    compile 'com.github.bumptech.glide:glide:3.6.0'
+    compile 'com.caverock:androidsvg:1.2.1'
```
License OK: [Apache License 2.0](https://github.com/BigBadaboom/androidsvg)

stain commented on app/build.gradle:
```diff
@@ -59,4 +61,7 @@ dependencies {
         exclude module: 'xpp3'
     }
+
+    compile 'com.github.bumptech.glide:glide:3.6.0'
```
License OK: [BSD 2-Clause and Apache](https://github.com/bumptech/glide/blob/master/LICENSE)

stain commented on app/build.gradle:
```diff
@@ -40,6 +40,8 @@ dependencies {
 
     compile 'com.squareup.retrofit2:retrofit:2.0.2'
     compile "com.squareup.retrofit2:adapter-rxjava:2.0.2"
+    compile 'com.squareup.okhttp3:logging-interceptor:3.2.0'
```
License OK: [Apache License 2.0](https://github.com/square/okhttp/blob/master/LICENSE.txt)

therajanmaurya commented:
@sagar15795 Please removes CheckStyle violation and add licenses and let me know for review.

sagar15795 commented:
@stain & @therajanmaurya 
I have removed the Checkstyle error . Also, I have added ASF  header in all modified or added file 

@stain 
I don't remember the URL of `placeholder.jpg` . So, I have replaced it with [Google Material icon of Photo](https://design.google.com/icons/#ic_photo)

therajanmaurya commented:
@stain Please look to licence and conform everything is Ok.
@sagar15795 this PR became cumbersome, So It needs time to review, tomorrow morning. I will review this PR.

Till that if want to work Make a branch from this branch and start implementing the detail Workflow rough Logic in Presenter and rough UI.
Tomorrow We will discuss about the Workflow details.

stain commented:
https://design.google.com/icons/ says "The icons are available under the Apache License Version 2.0. We'd love attribution in your app's "about" screen, but it's not required. The only thing we ask is that you not re-sell these icons."

"Apache License Version 2.0. " is fine, but "The only thing we ask is that you not re-sell these icons." is against the Apache License.. confusing.. 

Confusing license - https://github.com/google/material-design-icons/blob/master/LICENSE (and https://github.com/google/material-design-icons/#license ) says the license is [CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/), which DO require attribution.

google/material-design-icons/issues/417 is an issue about the license change.

>  Attribution — You must give appropriate credit, provide a link to the license, and indicate if changes were made. You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use.

but perhaps the README text "We'd love attribution in your app's about screen, but it's not required" does this mean you should then attribute it ELSEWHERE?

I've raised google/material-design-icons#424 to clarify the "do not sell" issue.

stain commented:
Thanks @sagar15795 - I don't see a problem with accepting the branch as it is now. (We can await the licensing question from Google and change the icon is needed).

@therajanmaurya do you think I'm OK to merge?

therajanmaurya commented:
@stain Please wait Actually I was not well today because of I did not review the PR. tomarrow morning I will reviewed then we can merge.

Thanks

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/remote/BaseApiManager.java:
```diff
@@ -26,6 +42,7 @@ public BaseApiManager() {
                 .baseUrl(ENDPOINT)
```
Replace ENDPOINT to END_POINT and make public static

therajanmaurya replied:
Replace ENDPOINT to END_POINT and make public static

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/DetailWorkflow.java:
```diff
@@ -0,0 +1,256 @@
+@Root(name = "workflow")
+public class DetailWorkflow implements Parcelable {
+
+    @Attribute(name = "resource", required = false)
```
Leave one line gape between the POJO objects.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/Author.java:
```diff
@@ -6,9 +24,7 @@
+
 public class Author implements Parcelable {
 
     public static final Parcelable.Creator<Author> CREATOR = new Parcelable.Creator<Author>() {
```
POJO objects should be first and then getter setter and at last Parcelable code.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/LicenseType.java:
```diff
@@ -0,0 +1,106 @@
+
+public class LicenseType implements Parcelable {
+
+    @Attribute(name = "resource", required = false)
```
Leave one line gap between the POJO objects.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/Tag.java:
```diff
@@ -0,0 +1,106 @@
+public class Tag implements Parcelable {
+
+    @Attribute(name = "resource", required = false)
+    String resource;
```
Leave one line gap between the POJO objects.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/Type.java:
```diff
@@ -0,0 +1,106 @@
+public class Type implements Parcelable {
+
+    @Attribute(name = "resource", required = false)
+    String resource;
```
Leave one line gap between the POJO objects.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/model/Uploader.java:
```diff
@@ -0,0 +1,105 @@
+
+@Root(name = "uploader")
+public class Uploader implements Parcelable {
+    @Attribute(name = "resource", required = false)
```
Leave one line gap between the POJO objects.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/remote/APIEndPoint.java:
```diff
@@ -0,0 +1,29 @@
+
+public class APIEndPoint {
+    public static final String ALLANNOUNCEMENT = "announcements.xml";
+    public static final String ANNOUNCEMENT = "announcement.xml";
```
Good Job 👍

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/remote/APIEndPoint.java:
```diff
@@ -0,0 +1,29 @@
+//This class contains all the Constants for API End Points
+
+public class APIEndPoint {
+    public static final String ALLANNOUNCEMENT = "announcements.xml";
```
ALLANNOUNCEMENT to ALL_ANNOUNCEMENT

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/remote/APIEndPoint.java:
```diff
@@ -0,0 +1,29 @@
+
+//This class contains all the Constants for API End Points
+
+public class APIEndPoint {
```
Good Job 👍  for this implementation

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/data/remote/APIEndPoint.java:
```diff
@@ -0,0 +1,29 @@
+public class APIEndPoint {
+    public static final String ALLANNOUNCEMENT = "announcements.xml";
+    public static final String ANNOUNCEMENT = "announcement.xml";
+    public static final String ALLWORKFLOW = "workflows.xml";
```
ALLWORKFLOW to ALL_WORKFLOW

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/adapter/WorkflowAdapter.java:
```diff
@@ -0,0 +1,149 @@
+
+    class ViewHolder extends RecyclerView.ViewHolder {
+
+        @BindView(R.id.tvDate)
```
Give a line gap between the Objects

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/adapter/EndlessRecyclerOnScrollListener.java:
```diff
@@ -28,7 +28,7 @@
     private boolean loading = true; // True if we are still waiting for the last set of data to
     // load.
-    private int visibleThreshold = 1; // The minimum amount of items to have below your current
+    private int visibleThreshold = 5; // The minimum amount of items to have below your current
```
initialize visibleThreshold = 1 because we don't want to API Request till that user will reach to the last element

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/workflow/WorkflowFragment.java:
```diff
@@ -0,0 +1,154 @@
+        mWorkflowAdapter = new WorkflowAdapter(mWorkflowList, getContext());
+        mRecyclerView.setAdapter(mWorkflowAdapter);
+
+        showProgressbar(true);
```
Remove this method call and Handle all Logic call in Presenter

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/workflow/WorkflowFragment.java:
```diff
@@ -0,0 +1,154 @@
+        showProgressbar(true);
+        mWorkflowPresenter.loadAllWorkflow(mPageNumber);
+
+        mRecyclerView.setOnScrollListener(new EndlessRecyclerOnScrollListener(layoutManager) {
```
setOnScrollListener is depreciated use addOnScrollListener

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/workflow/WorkflowFragment.java:
```diff
@@ -0,0 +1,154 @@
+                    mWorkflowAdapter.notifyItemInserted(mWorkflowList.size());
+                    ++mPageNumber;
+                    mWorkflowPresenter.loadAllWorkflow(mPageNumber);
+                    Log.i(LOG_TAG, "Loading more");
```
Please use Log carefully read this http://stackoverflow.com/a/7959379/3176277.
Here Lod.d should be used.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/workflow/WorkflowFragment.java:
```diff
@@ -0,0 +1,154 @@
+                    mWorkflowPresenter.loadAllWorkflow(mPageNumber);
+                    Log.i(LOG_TAG, "Loading more");
+                } else if (!mConnectionInfo.isConnectingToInternet()) {
+                    Log.i(LOG_TAG, "Internet not available. Not loading more posts.");
```
Same as above

therajanmaurya commented:
@sagar15795 I have looked to code and this needs some please look and let me know.

sagar15795 commented:
@therajanmaurya 

I have added all you input.
Have a look over code. Tell me if there is any suggestion or change . 

Thanks for in-person help.

therajanmaurya commented on app/src/main/java/org/apache/taverna/mobile/ui/adapter/WorkflowAdapter.java:
```diff
@@ -0,0 +1,159 @@
+    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        RecyclerView.ViewHolder vh;
+
+        requestBuilder = Glide.with(context)
```
Remove from here and put in Contractor

sagar15795 commented:
@therajanmaurya  

I have added your input.
Have a look over code.Tell me if there is any suggestion or change .

therajanmaurya commented:
Good Job 👍  @sagar15795 .

@stain Please Merge this PR

ianwdunlop commented:
There is some functionality missing/removed from the previous version. I don't think we can merge this yet. There appears to be no way to mark a workflow as a 'favourite'. You can no longer click on a workflow in the list and view it by itself. Having the workflow svg in the list view doesn't help a lot here since you can't see the whole thing since it is too small and appears to be zoomed in. In the favourites view (where I had previously adding one using the older codebase) clicking on the 'view' button pops up a small text box with some details where it should go to the actual workflow view. In the workflow list it is hard to read since the workflows sort of merge into each other. There needs to be a clearer break between them.

therajanmaurya commented:
1. Actually we do not want to add tue favourite workflow in workflows list. We want to add favourite in workflow details menu item. Where user can read the full workflow and mark as favourite or bookmark for later.
2. This PR have became very big with workflows list discussion. So I said to @sagar  make a sperate PR for workflow details. don't worry next PR will make full functionality of Workflows.
3. Yes here @sagar change the SVG to jpg! and set the full image in image view.
4. Favourite functionality will be added in next PR. Yes favourite click will be open workflow detail.

Actually on these topics. I wanna talk you all but you had busy schedule last week, So we thought to design by talk you later.

Thanks

sagar15795 commented:
@therajanmaurya  

As you said I have made changes and working on workflow details feature in sperate PR.

Please merge @ianwdunlop 

Thanks

sagar15795 commented:
@therajanmaurya  As you said I have made changes and working on workflow details feature in sperate PR.

Please merge @ianwdunlop 

Thanks

ianwdunlop commented:
There seem to be more and  more commits being pushed into this pull request. @sagar15795 you will need to tell us when to merge. It's great that you are adding all these things but it's getting too hard for me to follow all of this. Looking at the original comment in this thread there were a number of features being added/changed - are these still the things you are working on or have there been more? Anyway,  at some point you need to stop adding things here, get us to check it and then (assuming it is all ok and we have merged it) you can work on more features in another pull request.

therajanmaurya commented:
@sagar15795 I hope your Favorite logic is finished today. Please Let me know for review and 
Stop working on this. 

anything has any little bug or incompleteness Please Note down and we will do that in next PR

Thanks

sagar15795 commented:
I have done with all checked field. 

@therajanmaurya @ianwdunlop  
You can review it . 
Tell me if there is any suggestion or change .

I will add licence after your review.

Thanks .

therajanmaurya commented:
@sagar15795 some functionality is missing. we will add in next PR.
👍  Good Work.

@ianwdunlop Please Merge This PR. So @sagar15795 will finish this feature, because this PR is the base PR of this feature.

sagar15795 commented:
I have added ASF header in all modified or added file.

@ianwdunlop @stain  
Now you can merge it.
